In [1]:
import pandas as pd

In [3]:
df = pd.read_csv("/Users/jaygupta/Desktop/ITMD 522/Final Project/heart_disease_health_indicators_BRFSS2015.csv")

In [4]:
df

HeartDiseaseorAttack  HighBP  HighChol  CholCheck   BMI  Smoker  \
0                        0.0     1.0       1.0        1.0  40.0     1.0   
1                        0.0     0.0       0.0        0.0  25.0     1.0   
2                        0.0     1.0       1.0        1.0  28.0     0.0   
3                        0.0     1.0       0.0        1.0  27.0     0.0   
4                        0.0     1.0       1.0        1.0  24.0     0.0   
...                      ...     ...       ...        ...   ...     ...   
253675                   0.0     1.0       1.0        1.0  45.0     0.0   
253676                   0.0     1.0       1.0        1.0  18.0     0.0   
253677                   0.0     0.0       0.0        1.0  28.0     0.0   
253678                   0.0     1.0       0.0        1.0  23.0     0.0   
253679                   1.0     1.0       1.0        1.0  25.0     0.0   

        Stroke  Diabetes  PhysActivity  Fruits  ...  AnyHealthcare  \
0          0.0       0.0           0.0     0.0  ...            1.0   
1          0.0       0.0           1.0     0.0  ...            0.0   
2          0.0       0.0           0.0     1.0  ...            1.0   
3          0.0       0.0           1.0     1.0  ...            1.0   
4          0.0       0.0           1.0     1.0  ...            1.0   
...        ...       ...           ...     ...  ...            ...   
253675     0.0       0.0           0.0     1.0  ...            1.0   
253676     0.0       2.0           0.0     0.0  ...            1.0   
253677     0.0       0.0           1.0     1.0  ...            1.0   
253678     0.0       0.0           0.0     1.0  ...            1.0   
253679     0.0       2.0           1.0     1.0  ...            1.0   

        NoDocbcCost  GenHlth  MentHlth  PhysHlth  DiffWalk  Sex   Age  \
0               0.0      5.0      18.0      15.0       1.0  0.0   9.0   
1               1.0      3.0       0.0       0.0       0.0  0.0   7.0   
2               1.0      5.0      30.0      30.0       1.0  0.0   9.0   
3               0.0      2.0       0.0       0.0       0.0  0.0  11.0   
4               0.0      2.0       3.0       0.0       0.0  0.0  11.0   
...             ...      ...       ...       ...       ...  ...   ...   
253675          0.0      3.0       0.0       5.0       0.0  1.0   5.0   
253676          0.0      4.0       0.0       0.0       1.0  0.0  11.0   
253677          0.0      1.0       0.0       0.0       0.0  0.0   2.0   
253678          0.0      3.0       0.0       0.0       0.0  1.0   7.0   
253679          0.0      2.0       0.0       0.0       0.0  0.0   9.0   

        Education  Income  
0             4.0     3.0  
1             6.0     1.0  
2             4.0     8.0  
3             3.0     6.0  
4             5.0     4.0  
...           ...     ...  
253675        6.0     7.0  
253676        2.0     4.0  
253677        5.0     2.0  
253678        5.0     1.0  
253679        6.0     2.0  

[253680 rows x 22 columns]

#### As we can see, the data is usually made up of Integer and Float Values. These are Numerical attributes. Few of them are Discrete and few of them are continuous.

# Data PreProcessing 

## Checking for any missing values in the dataset

In [5]:
df.isna().sum()

HeartDiseaseorAttack    0
HighBP                  0
HighChol                0
CholCheck               0
BMI                     0
Smoker                  0
Stroke                  0
Diabetes                0
PhysActivity            0
Fruits                  0
Veggies                 0
HvyAlcoholConsump       0
AnyHealthcare           0
NoDocbcCost             0
GenHlth                 0
MentHlth                0
PhysHlth                0
DiffWalk                0
Sex                     0
Age                     0
Education               0
Income                  0
dtype: int64

###  There are no missing values in the data.

### As we are going to implement the classification task, we will scale the data into the range of 0-1

In [6]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler(feature_range=(0,1))
numeric_data = ['int64', 'float64']
cols_numeric = df.select_dtypes(include=numeric_data).columns.tolist()
cols_numeric_index=[df.columns.get_loc(col) for col in cols_numeric]
df[cols_numeric]=scaler.fit_transform(df[cols_numeric])
df = df[cols_numeric]
print(df[cols_numeric])

        HeartDiseaseorAttack  HighBP  HighChol  CholCheck       BMI  Smoker  \
0                        0.0     1.0       1.0        1.0  0.325581     1.0   
1                        0.0     0.0       0.0        0.0  0.151163     1.0   
2                        0.0     1.0       1.0        1.0  0.186047     0.0   
3                        0.0     1.0       0.0        1.0  0.174419     0.0   
4                        0.0     1.0       1.0        1.0  0.139535     0.0   
...                      ...     ...       ...        ...       ...     ...   
253675                   0.0     1.0       1.0        1.0  0.383721     0.0   
253676                   0.0     1.0       1.0        1.0  0.069767     0.0   
253677                   0.0     0.0       0.0        1.0  0.186047     0.0   
253678                   0.0     1.0       0.0        1.0  0.127907     0.0   
253679                   1.0     1.0       1.0        1.0  0.151163     0.0   

        Stroke  Diabetes  PhysActivity  Fruits  ...

### Further, we do not need to convert any attributes to numeric as all of them are numeric itself.

## Encoding the data for the Label 

In [7]:
# Label Encoding for Decision Tree. This process is also known as OneHot Encoding. 
from sklearn import preprocessing
y = df['HeartDiseaseorAttack']
le = preprocessing.LabelEncoder()
le.fit(y)
y_encoded = le.transform(y)
df['HeartDiseaseorAttack'] = y_encoded
x_features = df.drop('HeartDiseaseorAttack',1)
y_encoded

/var/folders/_m/t8lzzcn5137b7_h1b1mvd_bm0000gn/T/ipykernel_8958/1628918669.py:8: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  x_features = df.drop('HeartDiseaseorAttack',1)


array([0, 0, 0, ..., 0, 0, 1])

# Feature Selection

In [8]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

selector = SelectKBest(f_classif, k=10).fit(x_features, y_encoded)
print('\nSelected features by ANOVA:', selector.get_feature_names_out())


Selected features by ANOVA: ['HighBP' 'HighChol' 'Smoker' 'Stroke' 'Diabetes' 'GenHlth' 'PhysHlth'
 'DiffWalk' 'Age' 'Income']


## We have implemented feature selection with the help of ANOVA. The important factors or features that can cause a heart attack are HighBP, HighChol, Smoker, Stroke, Diabities, GenHealth,PhysHealth, DiffWalk, Age and Income.

### Creating a dataframe with the 5 top selected important features

In [9]:
df1 = df[["HighBP" , "HighChol", "Smoker", "Stroke", "Diabetes"]]

In [10]:
df1

HighBP  HighChol  Smoker  Stroke  Diabetes
0          1.0       1.0     1.0     0.0       0.0
1          0.0       0.0     1.0     0.0       0.0
2          1.0       1.0     0.0     0.0       0.0
3          1.0       0.0     0.0     0.0       0.0
4          1.0       1.0     0.0     0.0       0.0
...        ...       ...     ...     ...       ...
253675     1.0       1.0     0.0     0.0       0.0
253676     1.0       1.0     0.0     0.0       1.0
253677     0.0       0.0     0.0     0.0       0.0
253678     1.0       0.0     0.0     0.0       0.0
253679     1.0       1.0     0.0     0.0       1.0

[253680 rows x 5 columns]

## Splitting the Data with the help of Train Test Split.

In [11]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test =train_test_split(x_features, y_encoded, test_size=0.3)

In [12]:
x1_train, x1_test, y1_train, y1_test =train_test_split(df1, y_encoded, test_size=0.3)

# Implementing the classification Modules

### Model 1 -  Logistic Regression:

In [13]:
# Importing the neccessary packages
from sklearn import metrics 
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import make_scorer, precision_score, accuracy_score, recall_score
from sklearn.metrics import confusion_matrix as cm
from sklearn.metrics import f1_score

In [14]:
clf=LogisticRegression(penalty='l2',solver='lbfgs')
clf=clf.fit(x_train, y_train)
y_pred=clf.predict(x_test)

acc=accuracy_score(y_test, y_pred)
pre=precision_score(y_test, y_pred, average='macro')
f1=f1_score(y_test,y_pred, average='macro')
print('By hold-out evaluation: acc = ',acc, ', precision = ', pre, ', F1: ', f1)

By hold-out evaluation: acc =  0.9072190686429097 , precision =  0.7321853678691103 , F1:  0.5796090424801806


## HyperParameter Tuning for Logistic Regression to find out the best Model

In [16]:
clf=LogisticRegression(penalty='l1',solver='liblinear')
clf=clf.fit(x_train, y_train)
y_pred=clf.predict(x_test)

acc=accuracy_score(y_test, y_pred)
pre=precision_score(y_test, y_pred, average='micro')
f1=f1_score(y_test,y_pred, average='micro')
print('By hold-out evaluation: acc = ',acc, ', precision = ', pre, ', F1: ', f1)

By hold-out evaluation: acc =  0.9076921055397876 , precision =  0.9076921055397876 , F1:  0.9076921055397876


In [17]:
clf=LogisticRegression(penalty='l2',solver='newton-cg')
clf=clf.fit(x_train, y_train)
y_pred=clf.predict(x_test)

acc=accuracy_score(y_test, y_pred)
pre=precision_score(y_test, y_pred, average='macro')
f1=f1_score(y_test,y_pred, average='macro')
print('By hold-out evaluation: acc = ',acc, ', precision = ', pre, ', F1: ', f1)

By hold-out evaluation: acc =  0.9077315252811942 , precision =  0.7206670301000739 , F1:  0.5760879401802993


In [18]:
x_train, x_test, y_train, y_test = train_test_split(x_features, y_encoded, test_size=0.3)
clf=LogisticRegression(penalty='l1',solver='liblinear')
clf=clf.fit(x_train, y_train)
y_pred=clf.predict(x_test)

acc=accuracy_score(y_test, y_pred)
pre=precision_score(y_test, y_pred, average='micro')
f1=f1_score(y_test,y_pred, average='micro')
print('By hold-out evaluation: acc = ',acc, ', precision = ', pre, ', F1: ', f1)

By hold-out evaluation: acc =  0.9082702617470829 , precision =  0.9082702617470829 , F1:  0.9082702617470829


### Hence after evaluating the model using Logistic Regression, we found the accuracy to be 90.8% and the F1 score to be approximately 91%

## Implementing Logistic Regression with the best selected features

In [48]:
clf=LogisticRegression(penalty='l2',solver='lbfgs')
clf=clf.fit(x1_train, y1_train)
y_pred=clf.predict(x1_test)

acc=accuracy_score(y1_test, y_pred)
pre=precision_score(y1_test, y_pred, average='micro')
f1=f1_score(y1_test,y_pred, average='micro')
print('By hold-out evaluation: acc = ',acc, ', precision = ', pre, ', F1: ', f1)

By hold-out evaluation: acc =  0.906443813728582 , precision =  0.906443813728582 , F1:  0.906443813728582


### Thus, we can see that there is not much change in the accuracy and F1 Score with the help of the best selected features 

### Model 2 -  SVM:

In [14]:
from sklearn.svm import SVC
clf=SVC(kernel='linear', C=0.5, max_iter=1000)            
clf=clf.fit(x_train, y_train)
y_pred=clf.predict(x_test)
print('Accuracy: ', accuracy_score(y_test, y_pred), ', Precision: ', precision_score(y_test, y_pred, average='macro', ),
         ', Recall: ', recall_score(y_test, y_pred, average='macro', ), ', F1: ', f1_score(y_test, y_pred, average='micro'))

C:\Users\parth\anaconda3\lib\site-packages\sklearn\svm\_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=1000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


Accuracy:  0.10636760222852938 , Precision:  0.5275286687948701 , Recall:  0.5041897588312587 , F1:  0.10636760222852938


In [17]:
clf=SVC(kernel='rbf', C=1, max_iter=1500) 
clf=clf.fit(x_train, y_train)           
y_pred=clf.predict(x_test)
print('Accuracy: ', accuracy_score(y_test, y_pred), ', Precision: ', precision_score(y_test, y_pred, average='macro', ),
         ', Recall: ', recall_score(y_test, y_pred, average='macro', ), ', F1: ', f1_score(y_test, y_pred, average='micro'))

C:\Users\parth\anaconda3\lib\site-packages\sklearn\svm\_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=1500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


Accuracy:  0.8188137285819405 , Precision:  0.5427602021643181 , Recall:  0.5569750319446818 , F1:  0.8188137285819405


In [18]:
clf=SVC(kernel='linear', C=1.8, max_iter=8000)
clf=clf.fit(x_train, y_train)                 
y_pred=clf.predict(x_test)
print('Accuracy: ', accuracy_score(y_test, y_pred), ', Precision: ', precision_score(y_test, y_pred, average='macro', ),
         ', Recall: ', recall_score(y_test, y_pred, average='macro', ), ', F1: ', f1_score(y_test, y_pred, average='micro'))

C:\Users\parth\anaconda3\lib\site-packages\sklearn\svm\_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=8000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


Accuracy:  0.19631031220435194 , Precision:  0.469141768917804 , Recall:  0.4553667834061225 , F1:  0.1963103122043519


## Implementing SVM for the best features

In [17]:
clf=SVC(kernel='rbf', C=0.1 , max_iter=150) 
clf=clf.fit(x1_train, y1_train)           
y_pred=clf.predict(x1_test)
print('Accuracy: ', accuracy_score(y1_test, y_pred), ', Precision: ', precision_score(y1_test, y_pred, average='macro', ),
         ', Recall: ', recall_score(y1_test, y_pred, average='macro', ), ', F1: ', f1_score(y1_test, y_pred, average='micro'))

C:\Users\parth\anaconda3\lib\site-packages\sklearn\svm\_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=150).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


Accuracy:  0.5421134237359403 , Precision:  0.5292255654051202 , Recall:  0.5856169233099718 , F1:  0.5421134237359403


In [18]:
clf=SVC(kernel='linear', C=10 , max_iter=200) 
clf=clf.fit(x1_train, y1_train)           
y_pred=clf.predict(x1_test)
print('Accuracy: ', accuracy_score(y1_test, y_pred), ', Precision: ', precision_score(y1_test, y_pred, average='macro', ),
         ', Recall: ', recall_score(y1_test, y_pred, average='macro', ), ', F1: ', f1_score(y1_test, y_pred, average='micro'))

C:\Users\parth\anaconda3\lib\site-packages\sklearn\svm\_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=200).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


Accuracy:  0.094108062651109 , Precision:  0.0470540313255545 , Recall:  0.5 , F1:  0.094108062651109


C:\Users\parth\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


### Hence after evaluating the model using SVM, we found the accuracy to be 87.3% and the F1 score to be 87.3%

### Model 3 - KNN( K Nearest Neighbour)

In [15]:
from sklearn import neighbors
kvalues = [161,181,201,221,241,261]
for k in kvalues: 
    clf=neighbors.KNeighborsClassifier(k, metric='euclidean')
    clf.fit(x_train, y_train)
    y_pred = clf.predict(x_test)
    print('K =', k, ', Accuracy: ', accuracy_score(y_test, y_pred), ', Precision: ', precision_score(y_test, y_pred, average='macro', zero_division=0),
         ', Recall: ', recall_score(y_test, y_pred, average='macro', zero_division=0),', F1: ', f1_score(y1_test, y_pred, average='micro'))

K = 161 , Accuracy:  0.9065095132975928 , Precision:  0.7907054330901109 , Recall:  0.5142509168300826 , F1:  0.9020025228634501
K = 181 , Accuracy:  0.9065489330389993 , Precision:  0.8052441068775384 , Recall:  0.5131580324816392 , F1:  0.9023573005361085
K = 201 , Accuracy:  0.9064306738147798 , Precision:  0.8003154867158675 , Recall:  0.5125353674754125 , F1:  0.9025281194155367
K = 221 , Accuracy:  0.9065095132975928 , Precision:  0.8080814412561907 , Recall:  0.5125789233794152 , F1:  0.9026069588983496
K = 241 , Accuracy:  0.9064569536423841 , Precision:  0.8124079325172994 , Recall:  0.5118067779097835 , F1:  0.9028434773467886
K = 261 , Accuracy:  0.9064043939871754 , Precision:  0.8092094728983357 , Recall:  0.5115919635903741 , F1:  0.9028960370019973


## Thus, we can see that the accuracy of the Model is 90%. 

### Hyperparameter Tuning for KNN

In [58]:
from sklearn import neighbors
kvalues = [161,181,201,221,241,261]
for k in kvalues: 
    clf=neighbors.KNeighborsClassifier(k, metric='manhattan')
    clf.fit(x_train, y_train)
    y_pred = clf.predict(x_test)
    print('K =', k, ', Accuracy: ', accuracy_score(y_test, y_pred), ', Precision: ', precision_score(y_test, y_pred, average='macro', zero_division=0),
         ', Recall: ', recall_score(y_test, y_pred, average='macro', zero_division=0))

K = 161 , Accuracy:  0.9086644591611479 , Precision:  0.7699156747732305 , Recall:  0.5152115042268863
K = 181 , Accuracy:  0.9084805003679176 , Precision:  0.7640590141524192 , Recall:  0.5138992310614233
K = 201 , Accuracy:  0.9085987595921371 , Precision:  0.7717204079004768 , Recall:  0.5140918570045089
K = 221 , Accuracy:  0.9084410806265111 , Precision:  0.7683142715664688 , Recall:  0.512666599941168
K = 241 , Accuracy:  0.9085330600231263 , Precision:  0.7762102624983511 , Recall:  0.5126535491077969
K = 261 , Accuracy:  0.9083228214022916 , Precision:  0.7676270977837557 , Recall:  0.511263063435611


In [59]:
kvalues = [181,201,221]
for k in kvalues: 
    clf=neighbors.KNeighborsClassifier(k, metric='manhattan')
    clf.fit(x_train, y_train)
    y_pred = clf.predict(x_test)
    print('Accuracy: ', accuracy_score(y_test, y_pred), ', Precision: ', precision_score(y_test, y_pred, average='micro', ),
         ', Recall: ', recall_score(y_test, y_pred, average='micro', ), ', F1: ', f1_score(y_test, y_pred, average='macro'))

Accuracy:  0.9084805003679176 , Precision:  0.9084805003679176 , Recall:  0.9084805003679176 , F1:  0.5042620130377756
Accuracy:  0.9085987595921371 , Precision:  0.9085987595921371 , Recall:  0.9085987595921371 , F1:  0.5045841745232682
Accuracy:  0.9084410806265111 , Precision:  0.9084410806265111 , Recall:  0.9084410806265111 , F1:  0.5018065684804002


### Implementing KNN for Best Features

In [46]:
from sklearn import neighbors
kvalues = [161,181,201,221,241,261]
for k in kvalues: 
    clf=neighbors.KNeighborsClassifier(k, metric='euclidean')
    clf.fit(x1_train, y1_train)
    y_pred = clf.predict(x1_test)
    print('Accuracy: ', accuracy_score(y1_test, y_pred), ', Precision: ', 
          precision_score(y1_test, y_pred, average='micro', ),
         ', Recall: ', recall_score(y1_test, y_pred, average='micro', ),
          ', F1: ', f1_score(y1_test, y_pred, average='micro'))

Accuracy:  0.9065752128666036 , Precision:  0.9065752128666036 , Recall:  0.9065752128666036 , F1:  0.9065752128666036
Accuracy:  0.9065752128666036 , Precision:  0.9065752128666036 , Recall:  0.9065752128666036 , F1:  0.9065752128666036
Accuracy:  0.9065095132975928 , Precision:  0.9065095132975928 , Recall:  0.9065095132975928 , F1:  0.9065095132975928
Accuracy:  0.9066540523494165 , Precision:  0.9066540523494165 , Recall:  0.9066540523494165 , F1:  0.9066540523494165
Accuracy:  0.9065883527804057 , Precision:  0.9065883527804057 , Recall:  0.9065883527804057 , F1:  0.9065883527804057
Accuracy:  0.9065883527804057 , Precision:  0.9065883527804057 , Recall:  0.9065883527804057 , F1:  0.9065883527804057


In [47]:
from sklearn import neighbors
kvalues = [181,201,221]
for k in kvalues: 
    clf=neighbors.KNeighborsClassifier(k, metric='manhattan')
    clf.fit(x1_train, y1_train)
    y_pred = clf.predict(x1_test)
    print('Accuracy: ', accuracy_score(y1_test, y_pred), ', Precision: ', 
          precision_score(y1_test, y_pred, average='micro', ),
         ', Recall: ', recall_score(y1_test, y_pred, average='micro', )
          , ', F1: ', f1_score(y1_test, y_pred, average='micro'))

Accuracy:  0.9065752128666036 , Precision:  0.9065752128666036 , Recall:  0.9065752128666036 , F1:  0.9065752128666036
Accuracy:  0.9065095132975928 , Precision:  0.9065095132975928 , Recall:  0.9065095132975928 , F1:  0.9065095132975928
Accuracy:  0.9066540523494165 , Precision:  0.9066540523494165 , Recall:  0.9066540523494165 , F1:  0.9066540523494165


### Model 4 - Decision Tree

## Decision Tree Implementation 

In [32]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import make_scorer, accuracy_score, precision_score, recall_score, f1_score
clf=DecisionTreeClassifier(criterion='gini', max_depth=10, ccp_alpha = 0.1) 
clf=clf.fit(x_train, y_train)
y_pred=clf.predict(x_test)
acc=accuracy_score(y_pred, y_test)
pre=precision_score(y_pred, y_test)
rec=recall_score(y_pred, y_test)
print('Accuracy: ', accuracy_score(y_test, y_pred), ', Precision: ', precision_score(y_test, y_pred, average='micro', ),
         ', Recall: ', recall_score(y_test, y_pred, average='micro', ), ', F1: ', f1_score(y_test, y_pred, average='macro'))

Accuracy:  0.9052875013139914 , Precision:  0.9052875013139914 , Recall:  0.9052875013139914 , F1:  0.4751448275862069


/Users/jaygupta/opt/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


## Parameter Tuning to check for the accuracy of the Decision tree

In [32]:
clf=DecisionTreeClassifier(criterion='gini', ccp_alpha = 0.1 , max_features = 'log2' ,random_state = 1) 
clf=clf.fit(x_train, y_train)
y_pred=clf.predict(x_test)
acc=accuracy_score(y_pred, y_test)
pre=precision_score(y_pred, y_test)
rec=recall_score(y_pred, y_test)
print('Accuracy: ', accuracy_score(y_test, y_pred), ', Precision: ', precision_score(y_test, y_pred, average='macro', ),
         ', Recall: ', recall_score(y_test, y_pred, average='macro', ), ', F1: ', f1_score(y_test, y_pred, average='micro'))

Accuracy:  0.9057736781246715 , Precision:  0.45288683906233573 , Recall:  0.5 , F1:  0.9057736781246715


C:\Users\parth\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\parth\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [33]:
clf=DecisionTreeClassifier(criterion='entropy', max_depth=900000, ccp_alpha = 0.001, max_leaf_nodes = 200 )
clf=clf.fit(x_train, y_train)
y_pred=clf.predict(x_test)
acc=accuracy_score(y_pred, y_test)
pre=precision_score(y_pred, y_test)
rec=recall_score(y_pred, y_test)
print('Accuracy: ', accuracy_score(y_test, y_pred), ', Precision: ', precision_score(y_test, y_pred, average='macro', ),
         ', Recall: ', recall_score(y_test, y_pred, average='micro', ), ', F1: ', f1_score(y_test, y_pred, average='macro'))

Accuracy:  0.9059576369179018 , Precision:  0.7091197642176061 , Recall:  0.9059576369179018 , F1:  0.5383765979155641


### Constructing models for the decision trees using the best features

In [35]:
clf=DecisionTreeClassifier(criterion='gini', max_depth=1000000 , ccp_alpha = 0.001) 
clf=clf.fit(x1_train, y1_train)
y_pred=clf.predict(x1_test)
acc=accuracy_score(y_pred, y1_test)
pre=precision_score(y_pred, y1_test)
rec=recall_score(y_pred, y1_test)
print('Accuracy: ', accuracy_score(y1_test, y_pred),
      ', Precision: ', precision_score(y1_test, y_pred, average='micro', ),
         ', Recall: ', recall_score(y1_test, y_pred, average='micro', )
      , ', F1: ', f1_score(y1_test, y_pred, average='micro'))

Accuracy:  0.9059050772626932 , Precision:  0.9059050772626932 , Recall:  0.9059050772626932 , F1:  0.9059050772626932


/Users/jaygupta/opt/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [34]:
clf=DecisionTreeClassifier(criterion='entropy', max_depth=2500000 , ccp_alpha = 0.01 , random_state = 1)
clf=clf.fit(x1_train, y1_train)
y_pred=clf.predict(x1_test)
acc=accuracy_score(y_pred, y1_test)
pre=precision_score(y_pred, y1_test)
rec=recall_score(y_pred, y1_test)
print('Accuracy: ', accuracy_score(y1_test, y_pred), ', Precision: ', 
      precision_score(y1_test, y_pred, average='micro', ),
         ', Recall: ', recall_score(y1_test, y_pred, average='micro', ),
      ', F1: ', f1_score(y1_test, y_pred, average='micro'))

Accuracy:  0.9059050772626932 , Precision:  0.9059050772626932 , Recall:  0.9059050772626932 , F1:  0.9059050772626932


/Users/jaygupta/opt/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# Implementing Naive Bayes 

In [19]:
from sklearn.metrics import classification_report
from sklearn.naive_bayes import CategoricalNB,BernoulliNB,ComplementNB,MultinomialNB,GaussianNB
clf = GaussianNB()
clf.fit(x_train, y_train)
y_pred = clf.predict(x_test)
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='macro', )
recall = recall_score(y_test, y_pred, average='macro', )
f1score = f1_score(y_test, y_pred, average='micro')
print('Accuracy: ', accuracy, ', Precision: ', precision,
         ', Recall: ', recall, ', F1: ', f1score)

Accuracy:  0.8179464942709976 , Precision:  0.6075120491167205 , Recall:  0.6935335586527556 , F1:  0.8179464942709976


In [43]:
clf = ComplementNB()
clf.fit(x_train,y_train)
y_pred = clf.predict(x_test)
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='macro', )
recall = recall_score(y_test, y_pred, average='macro', )
f1score = f1_score(y_test, y_pred, average='micro')
print('Accuracy: ', accuracy, ', Precision: ', precision,
         ', Recall: ', recall, ', F1: ', f1score)

Accuracy:  0.7505781562072953 , Precision:  0.5863784165802413 , Recall:  0.7061681317517753 , F1:  0.7505781562072953


In [44]:
clf = MultinomialNB()
clf.fit(x_train,y_train)
y_pred = clf.predict(x_test)
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='macro', )
recall = recall_score(y_test, y_pred, average='macro', )
f1score = f1_score(y_test, y_pred, average='micro')
print('Accuracy: ', accuracy, ', Precision: ', precision,
         ', Recall: ', recall, ', F1: ', f1score)

Accuracy:  0.9025281194155367 , Precision:  0.6738363636094696 , Recall:  0.5683607410254761 , F1:  0.9025281194155367


In [45]:
clf = BernoulliNB()
clf.fit(x_train,y_train)
y_pred = clf.predict(x_test)
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='macro', )
recall = recall_score(y_test, y_pred, average='macro', )
f1score = f1_score(y_test, y_pred, average='micro')
print('Accuracy: ', accuracy, ', Precision: ', precision,
         ', Recall: ', recall, ', F1: ', f1score)

Accuracy:  0.8735283296541575 , Precision:  0.62984866433627 , Recall:  0.6365080687067644 , F1:  0.8735283296541575


In [47]:
clf = CategoricalNB()
clf.fit(x_train,y_train)
y_pred = clf.predict(x_test)
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='macro', )
recall = recall_score(y_test, y_pred, average='macro', )
f1score = f1_score(y_test, y_pred, average='micro')
print('Accuracy: ', accuracy, ', Precision: ', precision,
         ', Recall: ', recall, ', F1: ', f1score)

Accuracy:  0.8711631451697677 , Precision:  0.6302215730035967 , Recall:  0.6446371912000324 , F1:  0.8711631451697677


### Checking for the best features to check if there is any change in accuracy

In [37]:
clf = GaussianNB()
clf.fit(x1_train, y1_train)
y_pred = clf.predict(x1_test)
accuracy = accuracy_score(y1_test, y_pred)
precision = precision_score(y1_test, y_pred, average='micro', )
recall = recall_score(y1_test, y_pred, average='micro', )
f1score = f1_score(y1_test, y_pred, average='micro')
print('Accuracy: ', accuracy, ', Precision: ', precision,
         ', Recall: ', recall, ', F1: ', f1score)

Accuracy:  0.8805187637969095 , Precision:  0.8805187637969095 , Recall:  0.8805187637969095 , F1:  0.8805187637969094


In [38]:
clf = ComplementNB()
clf.fit(x1_train,y1_train)
y_pred = clf.predict(x1_test)
accuracy = accuracy_score(y1_test, y_pred)
precision = precision_score(y1_test, y_pred, average='micro', )
recall = recall_score(y1_test, y_pred, average='micro', )
f1score = f1_score(y1_test, y_pred, average='micro')
print('Accuracy: ', accuracy, ', Precision: ', precision,
         ', Recall: ', recall, ', F1: ', f1score)

Accuracy:  0.7455192893934616 , Precision:  0.7455192893934616 , Recall:  0.7455192893934616 , F1:  0.7455192893934616


In [39]:
clf = MultinomialNB()
clf.fit(x1_train,y1_train)
y_pred = clf.predict(x1_test)
accuracy = accuracy_score(y1_test, y_pred)
precision = precision_score(y1_test, y_pred, average='micro', )
recall = recall_score(y1_test, y_pred, average='micro', )
f1score = f1_score(y1_test, y_pred, average='micro')
print('Accuracy: ', accuracy, ', Precision: ', precision,
         ', Recall: ', recall, ', F1: ', f1score)

Accuracy:  0.9059050772626932 , Precision:  0.9059050772626932 , Recall:  0.9059050772626932 , F1:  0.9059050772626932


In [40]:
clf = BernoulliNB()
clf.fit(x1_train,y1_train)
y_pred = clf.predict(x1_test)
accuracy = accuracy_score(y1_test, y_pred)
precision = precision_score(y1_test, y_pred, average='micro', )
recall = recall_score(y1_test, y_pred, average='micro', )
f1score = f1_score(y1_test, y_pred, average='micro')
print('Accuracy: ', accuracy, ', Precision: ', precision,
         ', Recall: ', recall, ', F1: ', f1score)

Accuracy:  0.8859849679386104 , Precision:  0.8859849679386104 , Recall:  0.8859849679386104 , F1:  0.8859849679386104


In [41]:
clf = CategoricalNB()
clf.fit(x1_train,y1_train)
y_pred = clf.predict(x1_test)
accuracy = accuracy_score(y1_test, y_pred)
precision = precision_score(y1_test, y_pred, average='micro', )
recall = recall_score(y1_test, y_pred, average='micro', )
f1score = f1_score(y1_test, y_pred, average='micro')
print('Accuracy: ', accuracy, ', Precision: ', precision,
         ', Recall: ', recall, ', F1: ', f1score)

Accuracy:  0.8874566382844529 , Precision:  0.8874566382844529 , Recall:  0.8874566382844529 , F1:  0.8874566382844528


## Thus, we can see that the accuracy here has increased for the best features.

## Naive Bayes for best features

# Implementing Neural Networks

## To implement Nueral Networks, we need to check whether all the parameters are integers or float.  

In [42]:
x_features.dtypes

HighBP               float64
HighChol             float64
CholCheck            float64
BMI                  float64
Smoker               float64
Stroke               float64
Diabetes             float64
PhysActivity         float64
Fruits               float64
Veggies              float64
HvyAlcoholConsump    float64
AnyHealthcare        float64
NoDocbcCost          float64
GenHlth              float64
MentHlth             float64
PhysHlth             float64
DiffWalk             float64
Sex                  float64
Age                  float64
Education            float64
Income               float64
dtype: object

## Thus, implementing Nueral Networks.

In [15]:
from sklearn.neural_network import MLPClassifier

In [50]:
clf=MLPClassifier(solver='adam', alpha=1e-3,hidden_layer_sizes=(200,100,50,15), random_state=1)
clf=clf.fit(x_train, y_train)
y_pred=clf.predict(x_test)
acc=accuracy_score(y_test, y_pred)
pre=precision_score(y_test, y_pred, average='micro', )
rec=recall_score(y_test, y_pred, average='micro', )
f1score=f1_score(y_test, y_pred, average='micro')
print('Accuracy: ', acc, ', Precision: ', pre,
         ', Recall: ', rec, ', F1: ', f1score)

Accuracy:  0.891753390097761 , Precision:  0.891753390097761 , Recall:  0.891753390097761 , F1:  0.891753390097761


/Users/jaygupta/opt/anaconda3/lib/python3.9/site-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


## Parameter Tuning for Nueral Networks

In [51]:
clf=MLPClassifier(solver='sgd',activation = "relu",learning_rate_init = 0.0001,alpha=0.01,hidden_layer_sizes=(200))
clf=clf.fit(x_train, y_train)
y_pred=clf.predict(x_test)
acc=accuracy_score(y_test, y_pred)
pre=precision_score(y_test, y_pred, average='micro', )
rec=recall_score(y_test, y_pred, average='micro', )
f1score=f1_score(y_test, y_pred, average='micro')
print('Accuracy: ', acc, ', Precision: ', pre,
         ', Recall: ', rec, ', F1: ', f1score)

Accuracy:  0.9073898875223378 , Precision:  0.9073898875223378 , Recall:  0.9073898875223378 , F1:  0.9073898875223378


In [54]:
clf=MLPClassifier(solver='sgd', activation = "tanh", learning_rate_init = 0.0001,  alpha=0.01)
clf=clf.fit(x_train, y_train)
y_pred=clf.predict(x_test)
acc=accuracy_score(y_test, y_pred)
pre=precision_score(y_test, y_pred, average='micro', )
rec=recall_score(y_test, y_pred, average='micro', )
f1score=f1_score(y_test, y_pred, average='micro')
print('Accuracy: ', acc, ', Precision: ', pre,
         ', Recall: ', rec, ', F1: ', f1score)

Accuracy:  0.9073898875223378 , Precision:  0.9073898875223378 , Recall:  0.9073898875223378 , F1:  0.9073898875223378


### Implementing Nueral networks for the best features

In [17]:
clf=MLPClassifier(solver='adam', alpha=1e-3,hidden_layer_sizes=(200,100,50,15), random_state=1)
clf=clf.fit(x1_train, y1_train)
y_pred=clf.predict(x1_test)
acc=accuracy_score(y1_test, y_pred)
pre=precision_score(y1_test, y_pred, average='micro', )
rec=recall_score(y1_test, y_pred, average='micro', )
f1score=f1_score(y1_test, y_pred, average='micro')
print('Accuracy: ', acc, ', Precision: ', pre,
         ', Recall: ', rec, ', F1: ', f1score)

Accuracy:  0.9081388626090613 , Precision:  0.9081388626090613 , Recall:  0.9081388626090613 , F1:  0.9081388626090613


In [19]:
clf=MLPClassifier(solver='sgd', activation = "relu", learning_rate_init = 0.1,  alpha=1e-11)
clf=clf.fit(x_train, y_train)
y_pred=clf.predict(x_test)
acc=accuracy_score(y_test, y_pred)
pre=precision_score(y_test, y_pred, average='micro', )
rec=recall_score(y_test, y_pred, average='micro', )
f1score=f1_score(y_test, y_pred, average='micro')
print('Accuracy: ', acc, ', Precision: ', pre,
         ', Recall: ', rec, ', F1: ', f1score)

Accuracy:  0.9069168506254599 , Precision:  0.9069168506254599 , Recall:  0.9069168506254599 , F1:  0.9069168506254599
